plot/plot.go

```diff
@@ -0,0 +1,300 @@
+func (p *Plot) SetChartType(t ChartType) *Plot {
+	p.ChartType = t
+	return p
+}
+
+// GetTimeseries from a series Plot. Panics if Kind != SeriesKind.
+func (p *Plot) GetTimeseries() *stats.Timeseries {
+	if p.Kind != SeriesKind {
+		panic(errors.Reason("Kind is not SeriesKind"))
+	}
+	return stats.NewTimeseries().Init(p.Dates, p.Y)
+}
+
+// GetXY extracts X and Y slices from an untimed Plot. Panics if Kind != XYKind.
+func (p *Plot) GetXY() ([]float64, []float64) {
+	if p.Kind != XYKind {
+		panic(errors.Reason("Kind is not XYKind"))
+	}
+	return p.X, p.Y
+}
+
+// Graph containing Plots.
+type Graph struct {
+	Kind       Kind
+	Name       string
+	Title      string
+	XLabel     string
+	YLogScale  bool
+	PlotsRight []*Plot // plots using the right Y axis
+	PlotsLeft  []*Plot // plots using the left Y axis
+}
+
+func NewGraph(kind Kind, name string) *Graph {
+	return &Graph{
+		Kind:   kind,
+		Name:   name,
+		Title:  "Unnamed",
+		XLabel: "Value",
+	}
+}
+
+func (g *Graph) SetTitle(t string) *Graph {
+	g.Title = t
+	return g
+}
+
+func (g *Graph) SetXLabel(l string) *Graph {
+	g.XLabel = l
+	return g
+}
+
+func (g *Graph) SetYLogScale(b bool) *Graph {
+	g.YLogScale = b
+	return g
+}
+
+// AddPlotRight adds a plot to be displayed using the right Y axis. It's an
+// error if the plot and the Graph have different Kinds.
+func (g *Graph) AddPlotRight(p *Plot) error {
+	if p.Kind != g.Kind {
+		return errors.Reason("plot's kind [%s] != graph's kind [%s]",
+			p.Kind, g.Kind)
+	}
+	g.PlotsRight = append(g.PlotsRight, p)
+	return nil
+}
+
+// AddPlotLeft adds a plot to be displayed using the left Y axis. It's an error
+// if the plot and the Graph have different Kinds.
+func (g *Graph) AddPlotLeft(p *Plot) error {
+	if p.Kind != g.Kind {
+		return errors.Reason("plot's kind [%s] != graph's kind [%s]",
+			p.Kind, g.Kind)
+	}
+	g.PlotsLeft = append(g.PlotsLeft, p)
+	return nil
+}
+
+// Group of Graphs sharding the same X axis.
+type Group struct {
+	Kind      Kind              // must match Graph's Kind
+	Name      string            // internal unique name of the group
+	XLogScale bool              // whether to use log-scale for X axis
+	Graphs    []*Graph          // to preserve the order
+	graphMap  map[string]*Graph // for quick access
+}
+
+func NewGroup(kind Kind, name string) *Group {
+	return &Group{
+		Kind:     kind,
+		Name:     name,
+		graphMap: make(map[string]*Graph),
+	}
+}
+
+func (g *Group) SetXLogScale(b bool) *Group {
+	g.XLogScale = b
+	return g
+}
+
+// AddGraph to the Group. It's an error if the Graph Kind doesn't match the
+// Group Kind.
+func (g *Group) AddGraph(graph *Graph) error {
+	if g.Kind != graph.Kind {
+		return errors.Reason("group's Kind [%s] != graph's Kind [%s]",
+			g.Kind, graph.Kind)
+	}
+	if _, ok := g.graphMap[graph.Name]; ok {
+		return errors.Reason("graph %s already exists in group %s",
+			graph.Name, g.Name)
+	}
+	g.Graphs = append(g.Graphs, graph)
+	g.graphMap[graph.Name] = graph
```

Comment:
Factor this out into a helper, to ensure we always keep these two values in sync.